Azimuthal Integration Refactor

CSSFrancis opened the pull request:
---
**Name:** Azimuthal Integration Refactor
**About:** This pull request cleans up the Azimuthal integration methods for the diffraction2d class

---

**Release Notes**
> new feature / improvement / bugfix 
> Summary: Simplifies Azimuthal Integration and  adds returning summed integration vs mean integration 

**What does this PR do? Please describe and/or link to an open issue.**
This PR Introduces a slightly different work flow to the Azimuthual integration.  Namely it removes the "Slow Integration method" in favor of only creating one Azimuthal Inegrator Object in PyFAI(what was perviously the fast method).  Due to the large overhead of creating multiple AI objects this method is almost unusable due to it's speed.

This does mean that a user cannot define multiple affine corrections (Might be an issue but it most cases this should be constant for some data set?), and has to use the center direct beam method if they want to use multiple centers for the integration.

I've also added a sum option which gives the Radial integration rather than the Azimuthal integration. This could potentially be a separate method. 


**Describe alternatives you've considered**
- We could keep things as they are now, but I think that this helps us to direct our users towards faster implementations.
- Only define a slow method for using multiple affine corrections

**Are there any known issues? Do you need help?**
I might have to look into the Sum that is being calculated and make sure that it still accounts for the mask.

dnjohnstone commented:
Thanks for this, can you resolve the merge conflicts please @CSSFrancis ?

pc494 commented:
This looks like a big improvement, feel free to ping me for a review once the merge conflicts are resolved (hopefully an easy task)

CSSFrancis commented:
Hey I was wondering what the proper thing to do when passing multiple **kwargs type arguments to some function. 

The way I see it there are two options:

- Use a dictionary to hold all of the values or set it to None as a default. (I'm not sure why map_dict= None is preferable to map_dict={})
- We could just allow for **kwargs to be added and then separate them out 
` scatter_args = [k for k, v in inspect.signature(hs.signals.Signal2D.map).parameters.items()]` 

It means that you can't have the same keyword argument in two functions but it allows for the user to not have to define separate dictionaries (or dictionaries at all).

magnunor commented:
> has to use the center direct beam method if they want to use multiple centers for the integration

I'm guessing this means it isn't possible to input an array of different centre positions, with each azimuthal integration having a different centre?

For datasets which can fit into memory, this should be fine. However for really big datasets using packages such as `dask`, we ideally don't want to shift the data across the different chunks, as this can add a lot of overhead in the processing. This is how I've been doing it in the implementation from pixstem: first find the centre position using something like centre of mass in the data chunks in the centre, then do an azimuthal integration across all the probe positions. All in `dask`, so even really, really big datasets can be "reduced" to 3 dimensions data for exploration (x, y, scattering angle), even if the centre beam moves around a lot.

CSSFrancis commented:
> I'm guessing this means it isn't possible to input an array of different centre positions, with each azimuthal integration having a different centre?

So there are a couple of different options here.  What I was considering was a workflow which first you center the dataset using something like the diffraction2d.center_direct_beam() method (basically just the hyperspy align2d method) and then you perform the integration. I'm not certain about how that method scales with size/ lazy implementations. I could see it being a problem with dask if it isn't properly handled.

> For datasets which can fit into memory, this should be fine. However for really big datasets using packages such as `dask`, we ideally don't want to shift the data across the different chunks, as this can add a lot of overhead in the processing. This is how I've been doing it in the implementation from pixstem: first find the centre position using something like centre of mass in the data chunks in the centre, then do an azimuthal integration across all the probe positions. All in `dask`, so even really, really big datasets can be "reduced" to 3 dimensions data for exploration (x, y, scattering angle), even if the centre beam moves around a lot.

We are kind of limited by PyFAI to some extents.  Part of what makes it fast is that fact that it precalculates many of the spatial bins based on things like a spline interpolation, the center of the diffraction points and the Ewald sphere.  From my understanding changing the center point means changing the entire set up Azimuthal Integrator and that means a huge slow down.  

Part of me thinks it might be possible to shift the Azimuthal Inetegrator center but that might involve hacking the PyFAI  implementation a bit more.

I guess the other thing to think about is if there is ever a time where you need multiple affine corrections?  From my experience I would say that would be rare but maybe I am wrong.

magnunor commented:
> So there are a couple of different options here. What I was considering was a workflow which first you center the dataset using something like the diffraction2d.center_direct_beam() method (basically just the hyperspy align2d method) and then you perform the integration. I'm not certain about how that method scales with size/ lazy implementations. I could see it being a problem with dask if it isn't properly handled.

Implementing a lazy/dask version might be possible actually, but I haven't had the time to test how it affects the processing times for big datasets. I don't think this pull request need to handle multiple center positions, as it is (currently) a small minority of cases which can't use the `center_direct_beam()` method.

> I guess the other thing to think about is if there is ever a time where you need multiple affine corrections? From my experience I would say that would be rare but maybe I am wrong.

For some datasets with a very large field of view, there might be a need for this. But I also think this is for a very small percentage of datasets, so no need to take that into account for this pull request.

CSSFrancis commented:
Writing this out I was considering how the calibration workflow should go.  We could integrate more of the PyFAI functionality/ standards such as outputting a .poni file from the calibration generator and give the diffraction2d class a pyfai.AzimuthalIntegrator attribute. 

From that point we could implement every one of the azimuthal integrator functions as one line once the diffraction2d.ai attribute is initialized.  We still run into the shifting center point issue, but I think it outsources some of the current issues with the calibration pipeline while giving access to the calibration pipelines already implemented in PyFAI

pc494 commented:
When we've worked with crystallographic (spot pattern) data in the past, we've set up all our workflows in such a way that the patterns have to be centred first. From what I understand of the above that would also be an option here, and I think it would be my preference.

I can't speak to the calibration workflow. In terms of **kwargs, there isn't (as you probably found when you googled it) a "pythonic" way of doing it [1], and I would approach it on a case by case basis. Hyperspy uses dictionaries :`navigator_kwrgs` springs to mind, for this purpose in some places.

[1] - https://stackoverflow.com/questions/26534134/python-pass-different-kwargs-to-multiple-functions

CSSFrancis commented:
So this is kind of backwards but I rewrote the Azimuthal Integration methods with the idea of how to change our calibration pipeline.  In following the AzimuthalIntegrator class in pyFAI I think that we should almost Duck-type the Diffraction2D class so that it operates like an AzimuthalIntgrator object in pyFAI.

What I have done is:
- Added the units k_nm^-1 and k_A^-1 to the list of pyFAI units which reduces the need to have "pyxem" units as an option
- added a .ai property to Diffraction2D which needs to initialized either from the units of the diffraction2D object or from some AzimuthalIntegrator object. 

This sets up  a couple of new things we can add:
- Remake the calibration pipeline to create a AzimuthalIntegrator object
- Save calibrations in a .poni file that can be used in pyFAI as well
- support MultipleGeometry setups
- Easy implementation of all Azimuthal Integrator methods (Integrate1d, Integrate2d, Integrate_radially, sigma_clip, seperate, etc.)
- Access to pyFAI calibration methods.

CSSFrancis commented:
I added in some of the other integration methods from pyFAI. I am not sure how useful they are.  There is a separate method which separates Amorphous from Bragg diffraction which I would like to implement but that is a little more difficult to implement outside of the "map" function.

dnjohnstone commented:
Thanks for this @CSSFrancis - I like the idea of using more of the pyFAI functionality in the calibration steps and outputting a .poni file to use here. I think this might also fit well with some of the XRD stuff @ptim0626 and @pquinn-dls wanted in terms of calibration.

In terms of the scope of this PR though - things looked pretty good initially but are we expanding a bit too much. Could we get the core changes in to this one, then look at the calibration pipeline in the other PR you opened, and then come back to this?

tjof2 commented:
@CSSFrancis

> I'm not sure why map_dict= None is preferable to map_dict={}

See e.g. https://florimond.dev/blog/articles/2018/08/python-mutable-defaults-are-the-source-of-all-evil/ for details on mutable defaults such as `{}`
